hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/access/TombstoneAccessDelegate.java

```diff
@@ -0,0 +1,175 @@
+		}
+		else if (value instanceof FutureUpdate) {
+			return ((FutureUpdate) value).getValue();
+		}
+		else {
+			return value;
+		}
+	}
+
+	@Override
+	public boolean putFromLoad(SessionImplementor session, Object key, Object value, long txTimestamp, Object version) {
+		return putFromLoad(session, key, value, txTimestamp, version, false);
+	}
+
+	@Override
+	public boolean putFromLoad(SessionImplementor session, Object key, Object value, long txTimestamp, Object version, boolean minimalPutOverride) throws CacheException {
+		long lastRegionInvalidation = region.getLastRegionInvalidation();
+		if (txTimestamp < lastRegionInvalidation) {
+			log.tracef("putFromLoad not executed since tx started at %d, before last region invalidation finished = %d", txTimestamp, lastRegionInvalidation);
+			return false;
+		}
+		if (minimalPutOverride) {
+			Object prev = cache.get(key);
+			if (prev instanceof Tombstone) {
+				Tombstone tombstone = (Tombstone) prev;
+				long lastTimestamp = tombstone.getLastTimestamp();
+				if (txTimestamp <= lastTimestamp) {
+					log.tracef("putFromLoad not executed since tx started at %d, before last invalidation finished = %d", txTimestamp, lastTimestamp);
+					return false;
+				}
+			}
+			else if (prev != null) {
+				log.tracef("putFromLoad not executed since cache contains %s", prev);
+				return false;
+			}
+		}
+		// we can't use putForExternalRead since the PFER flag means that entry is not wrapped into context
+		// when it is present in the container. TombstoneCallInterceptor will deal with this.
+		putFromLoadCache.put(key, new TombstoneUpdate(session.getTimestamp(), value));
+		return true;
+	}
+
+	@Override
+	public boolean insert(SessionImplementor session, Object key, Object value, Object version) throws CacheException {
+		write(session, key, value);
+		return true;
+	}
+
+	@Override
+	public boolean update(SessionImplementor session, Object key, Object value, Object currentVersion, Object previousVersion) throws CacheException {
+		write(session, key, value);
+		return true;
+	}
+
+	protected void write(SessionImplementor session, Object key, Object value) {
+		TransactionCoordinator tc = session.getTransactionCoordinator();
+		FutureUpdateSynchronization sync = new FutureUpdateSynchronization(tc, writeCache, requiresTransaction, key, value);
+		// ReservedEntry is handled in TombstoneCallInterceptor
+		writeCache.put(key, new FutureUpdate(sync.getUuid(), null), region.getTombstoneExpiration(), TimeUnit.MILLISECONDS);
+		tc.getLocalSynchronizations().registerSynchronization(sync);
+	}
+
+	@Override
+	public void remove(SessionImplementor session, Object key) throws CacheException {
+		TransactionCoordinator transactionCoordinator = session.getTransactionCoordinator();
+		TombstoneSynchronization sync = new TombstoneSynchronization(transactionCoordinator, asyncWriteCache, requiresTransaction, region, key);
+		transactionCoordinator.getLocalSynchronizations().registerSynchronization(sync);
```

Comment:
No, and actually the synchronization should be registered only after successful put(), because we don't want to run the synchronization if it fails -> we're probably going to rollback the transaction. Well spotted!